src/main/java/com/wavefront/sdk/common/clients/WavefrontMultiClient.java

```diff
@@ -0,0 +1,145 @@
+/**
+ * WavefrontMultiClient supports multicasting metrics/distributions/spans
+ * onto multiple endpoints for either Proxy or Direct Ingestion.
+ * User should probably attempt to reconnect when exceptions are thrown from any methods.
+ *
+ * @author Mike McMahon (mike.mcmahon@wavefront.com).
+ */
+public class WavefrontMultiClient<T extends WavefrontSender & Runnable> implements WavefrontSender, Runnable {
+    private static final Logger logger = Logger.getLogger(
+            WavefrontProxyClient.class.getCanonicalName());
+
+    private final ConcurrentHashMap<String, T> wavefrontSenders = new ConcurrentHashMap<>();
+
+    public static class Builder<T extends WavefrontSender & Runnable> {
+        private final ConcurrentHashMap<String, T> wavefrontSenders = new ConcurrentHashMap<>();
+
+        public Builder withWavefrontSender(T sender) {
+            if (wavefrontSenders.containsKey(sender.getClientId()))
+                throw new RuntimeException("Duplicate id specified");
+
+            wavefrontSenders.put(sender.getClientId(), sender);
+            return this;
+        }
+
+        public WavefrontMultiClient<T> build() {
+            return new WavefrontMultiClient<T>(this);
+        }
+    }
+
+    private WavefrontMultiClient(Builder<T> builder) {
+        this.wavefrontSenders.putAll(builder.wavefrontSenders);
+    }
+
+    /**
+     * Provide direct access to a specific client by id.
+     * @param clientId the unique client ID generated when a Proxy or DirectIngestion Client is instantiated.
+     * @return The client found or null if no client is found matching the supplied clientId.
+     */
+    public T getClient(String clientId) {
+        return wavefrontSenders.getOrDefault(clientId, null);
+    }
+
+    @Override
+    public void flush() throws IOException {
+        for (T client : wavefrontSenders.values()) {
+            client.flush();
+        }
+    }
+
+    @Override
+    public int getFailureCount() {
+        int failureCount = 0;
+        for (T client : wavefrontSenders.values()) {
+            failureCount += client.getFailureCount();
+        }
+        return failureCount;
+    }
+
+    /**
+     * Obtain the failure counts per endpoint.
+     * @return a map of clientId's to failures per endpoint.
+     */
+    public Map<String, Integer> getFailureCountPerSender() {
+        final ConcurrentHashMap<String, Integer> failuresPerSender = new ConcurrentHashMap<>();
+        for (Map.Entry<String, T> e : wavefrontSenders.entrySet()) {
+            failuresPerSender.put(e.getKey(), e.getValue().getFailureCount());
+        }
+
+        return failuresPerSender;
+    }
+
+    @Override
+    public void sendMetric(String name, double value, @Nullable Long timestamp,
+                           @Nullable String source, @Nullable Map<String, String> tags)
+            throws IOException {
+        for (T client : wavefrontSenders.values()) {
+            client.sendMetric(name, value, timestamp, source, tags);
```

Comment:
For all of the send methods and flush, do we want a single proxy failure to bail and return or attempt to send to other proxies and bubble up a composite error?

Reply:
Hm that is a good point, it may be beneficial to specifically state which one failed so that a retry against that particular client may be made, and still attempt against others.   what do you envision for a composite error here?

Reply:
Don't need to expand the scope of this PR. Can be addressed in future improvement PRs as you find the time. We can do a few things:

1. Proxy/Direct clients emit a lot of internal metrics: https://github.com/wavefrontHQ/wavefront-sdk-java/blob/06d50d8142ef9b7a93da95cf9565d660e0078a44/src/main/java/com/wavefront/sdk/proxy/WavefrontProxyClient.java#L65

We can try to emit internal metrics with the specific client id to distinguish errors.

2. We can introduce a composite exception that extends an IOException. something like this: http://baddotrobot.com/blog/2011/08/29/catching-multiple-exceptions-and/